arangod/Aql/ExecutionBlockImpl.h

```diff
@@ -360,31 +361,43 @@ class ExecutionBlockImpl final : public ExecutionBlock {
    * consumed.
    */
   struct PrefetchTask {
-    enum class State { Pending, InProgress, Finished, Consumed };
+    enum class Status { Pending, InProgress, Finished, Consumed };
     using PrefetchResult =
         std::tuple<ExecutionState, SkipResult, typename Fetcher::DataRange>;
 
+    explicit PrefetchTask(ExecutionBlockImpl& block, AqlCallStack const& stack)
+        : _block(block), _stack(stack) {}
+
     bool isConsumed() const noexcept;
     bool tryClaim() noexcept;
+    bool tryClaimOrAbandon() noexcept;
     void waitFor() const noexcept;
-    void reset() noexcept;
+    bool rearmForNextCall(AqlCallStack const& stack);
     void discard(bool isFinished) noexcept;
     // note: this will throw if the PrefetchTask ran into an exception
     // during execution.
     PrefetchResult stealResult();
     void wakeupWaiter() noexcept;
     void setFailure(Result&& result);
 
-    void execute(ExecutionBlockImpl& block, AqlCallStack& stack);
+    void execute();
 
    private:
-    std::atomic<State> _state{State::Pending};
+    void updateStatus(Status status, std::memory_order memoryOrder) noexcept;
+    struct State {
+      Status status;
+      bool abandoned;
+    };
+    std::atomic<State> _state{{Status::Pending, false}};
```

Comment:
I verified that with all relevant compilers (g++-11, g++-13, clang++-16) this atomic is always lock-free (i.e. `std::atomic<State>::is_always_lock_free == true`). Shall we also add an assertion to have this guaranteed during the compile process?

Reply:
I figured it should be lock-free since this is a POD type of (probably) size 2, but yes, let's add an assertion for that!